### PR TITLE
Create API service for MVP

### LINKED ISSUE #17: Create API service for MVP
### Description
We need to be able to retrieve data from an external resource. This should (probably) be handled by an API service that is injected in the service setup. 

Take a look at the current setup, and try to understand how `AddSingleton` and `AddScoped` works.

Preferably document this in the Wiki as well, as this serves as the foundation for the DI (Dependency Injection).

### Subtasks
- [x] Assign project, milestone, labels and assignees
- [x] Create API service
- [x] Add API service to DI chain (services)
- [ ] Migrate dummy data source to met.api.no

### Related sources
- https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-5.0
- Partially blocked by #16 (we should only type the response type once)
- Blocked by #15

### PR DESCRIPTION
**Pull request by FredrikAugust**
- Create POCO and GraphQL types for data objects
- Wire up HTTP Client
- Send REST requests to api.met.no

Collaborative effort!

Closes #17

### Subtasks
- [x] Add support for decimals as input for coordinates
- [x] Create descriptions for fields in the GraphQL query
- [x] Fix sider linting issues

### CODE REVIEW

**Comment by fredrikbw** [comment]
Now wired up to both the Todo-API and the api.met.no 

- [ ]  Add support for decimals as input for coordinates
- [ ] Create descriptions for fields in the GraphQL query

**Comment by FredrikAugust** [comment]
Sider coming in strong with 19 issues 🥵

**Comment by fredrikbw** [comment]
They are all from the Forecast.cs file which has been autogenerated, so it's propably an easy fix!

**Comment by fredrikbw** [comment]
Warnings regarding nesting requires the Forecast.cs file to be split up into separate files, which isn't necessarily a good idea.

**Comment by FredrikAugust** [comment]
I see. You could look for a flag to ignore the file in the documentation.
Some of the warnings are sound, however, and should probably be fixed if not too hard to implement.

**Comment by fredrikbw** [comment]
All hail Lord Sider 🙏

File: API/MetAPI/Forecast.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Text.Json.Serialization;
+using Backend.API.MetAPI.SubClasses;
+
+namespace Backend.API.MetAPI
+{
+    public class Forecast
```
Comment (Kpaubert) [inline API/MetAPI/Forecast.cs:6]: Shouldn't this file be in the SubClass directory?

File: Startup.cs [hunk 2]
```diff
@@ -26,19 +27,28 @@ public Startup(IConfiguration configuration, IWebHostEnvironment environment)
         public void ConfigureServices(IServiceCollection services)
         {
             services
-                .AddSingleton<SampleSchema>()
+                .AddSingleton<IDataRetrievalService, DataRetrievalService>()
+                .AddSingleton<IForecastDataRetrievalService, ForecastDataRetrievalService>()
```
Comment (FredrikAugust) [inline Startup.cs:31]: We should only have one data retrieval service as HTTP Client should only be instantiated once. You may either move the functions in to the `DataRetrivalService` or create a subclass in said service and instantiate it in the constructor of the service.

File: Startup.cs [hunk 3]
```diff
@@ -26,19 +27,28 @@ public Startup(IConfiguration configuration, IWebHostEnvironment environment)
...
+                .AddSingleton<TodoSchema>()
+                .AddSingleton<TodoQuery>()
+                .AddSingleton<ForecastSchema>()
+                .AddSingleton<ForecastQuery>()
+                .AddHttpContextAccessor()
```
Comment (FredrikAugust) [inline Startup.cs:36]: What does this do?

File: Startup.cs [hunk 4]
```diff
@@ -47,12 +57,43 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+
+                PrettierPrintWidth = 80,
+                PrettierTabWidth = 2,
+                PrettierUseTabs = true,
+
+                SchemaDisableComments = false,
+                SchemaPollingEnabled = true,
+                SchemaPollingEndpointFilter = "*localhost*",
+                SchemaPollingInterval = 5000
+            });
```
Comment (FredrikAugust) [inline Startup.cs:93]: I believe we should only have one GraphQL endpoint. The client shouldn't have to make multiple connections to retrieve different data sources.

File: Startup.cs [hunk 5]
```diff
@@ -47,12 +57,43 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+            app.UseGraphQLWebSockets<ForecastSchema>();
+            app.UseGraphQLWebSockets<TodoSchema>();
 
-            app.UseGraphQL<SampleSchema, GraphQLHttpMiddleware<SampleSchema>>();
+            /*
+             * This will create two endpoints, based on https://github.com/graphql-dotnet/examples/blob/master/src/AspNetCoreMulti/Example/Startup.cs
+             */
+            app.UseGraphQLPlayground(new GraphQLPlaygroundOptions
+            {
+                GraphQLEndPoint = "/api/forecast",
```
Comment (FredrikAugust) [inline Startup.cs:74]: This can then be reverted to the original value (`/graphql`)

File: Startup.cs [hunk 6]
```diff
@@ -47,12 +57,43 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+                PrettierPrintWidth = 80,
+                PrettierTabWidth = 2,
+                PrettierUseTabs = true,
+
+                SchemaDisableComments = false,
+                SchemaPollingEnabled = true,
+                SchemaPollingEndpointFilter = "*localhost*",
+                SchemaPollingInterval = 5000
+            });
             app.UseGraphQLPlayground(new GraphQLPlaygroundOptions
```
Comment (FredrikAugust) [inline Startup.cs:94]: This block can be deleted when moving to one graphql endpoint.

File: Startup.cs [hunk 7]
```diff
@@ -47,12 +57,43 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                 app.UseDeveloperExceptionPage();
 
             app.UseWebSockets();
-            app.UseGraphQLWebSockets<SampleSchema>();
+            // Replaced the HttpMiddleware with paths for each schema
+           // app.UseGraphQL<ForecastSchema, GraphQLHttpMiddleware<ForecastSchema>>();
+            app.UseGraphQL<ForecastSchema>("/api/forecast");
+            //app.UseGraphQL<TodoSchema, GraphQLHttpMiddleware<TodoSchema>>();
+            app.UseGraphQL<TodoSchema>("/api/todo");
```
Comment (FredrikAugust) [inline Startup.cs:64]: Ref comment about one schema.

File: Startup.cs [hunk 8]
```diff
@@ -47,12 +57,43 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+
+            app.UseGraphQLWebSockets<ForecastSchema>();
+            app.UseGraphQLWebSockets<TodoSchema>();
```
Comment (FredrikAugust) [inline Startup.cs:67]: Previous comment

File: Startup.cs [hunk 9]
```diff
@@ -26,19 +27,28 @@ public Startup(IConfiguration configuration, IWebHostEnvironment environment)
...
                         logger.LogError("{Error} occured", ctx.OriginalException.Message);
                 })
+
                 .AddSystemTextJson(deserializerSettings => { }, serializerSettings => { })
                 .AddErrorInfoProvider(opt => opt.ExposeExceptionStackTrace = Environment.IsDevelopment())
                 .AddWebSockets()
                 .AddDataLoader()
-                .AddGraphTypes(typeof(SampleSchema));
+                .AddGraphTypes(typeof(ForecastSchema))
+                .AddGraphTypes(typeof(TodoSchema));
```
Comment (FredrikAugust) [inline Startup.cs:51]: Ditto

File: Backend.sln.DotSettings.user [hunk 10]
```diff
@@ -1,7 +1,7 @@
...
+  &lt;Project Location="\Users\fmalmo\Code\github.com\IT2901-SINTEF01\Backend\BackendTests" Presentation="&amp;lt;BackendTests&amp;gt;" /&gt;&#xD;
 &lt;/SessionState&gt;</s:String>
-	<s:String x:Key="/Default/Environment/UnitTesting/UnitTestSessionStore/Sessions/=f59c5c36_002D0f6d_002D4b69_002Dac92_002D71d6d1451f2f/@EntryIndexedValue">&lt;SessionState ContinuousTestingMode="0" IsActive="True" Name="All tests from Solution" xmlns="urn:schemas-jetbrains-com:jetbrains-ut-session"&gt;
-  &lt;Solution /&gt;
+	<s:String x:Key="/Default/Environment/UnitTesting/UnitTestSessionStore/Sessions/=f59c5c36_002D0f6d_002D4b69_002Dac92_002D71d6d1451f2f/@EntryIndexedValue">&lt;SessionState ContinuousTestingMode="0" IsActive="True" Name="All tests from Solution" xmlns="urn:schemas-jetbrains-com:jetbrains-ut-session"&gt;&#xD;
+  &lt;Solution /&gt;&#xD;
 &lt;/SessionState&gt;</s:String></wpf:ResourceDictionary>
```
Comment (FredrikAugust) [inline Backend.sln.DotSettings.user:7]: This file should be added to `.gitignore` and deleted. My bad, forgot to do it before handing the branch off to you:)

File: API/Services/ForecastDataRetrievalService.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        public ForecastDataRetrievalService()
+        {
+            this.HttpClient = new HttpClient();
+            this.HttpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            this.HttpClient.DefaultRequestHeaders.UserAgent.ParseAdd(
+                "Mozilla/5.0 (Macintosh; Intel Mac OS X x.y; rv:42.0) Gecko/20100101 Firefox/42.0");
+        }
+        public async Task<Forecast> GetForecast(float lat, float lon)
+        {
+            // todo: Add support for decimals to be used, also see ForecastQuery.cs
```
Comment (FredrikAugust) [inline API/Services/ForecastDataRetrievalService.cs:31]: Create issue for things that should be fixed instead of adding `TODO`s to the code. Makes it a bit easier to keep track of.

File: API/Services/ForecastDataRetrievalService.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+using System.Threading.Tasks;
+using Backend.API.MetAPI;
+
+namespace Backend.API.Services
+{
+    public interface IForecastDataRetrievalService
+    {
+        //HttpClient HttpClient { get; init; }
```
Comment (FredrikAugust) [inline API/Services/ForecastDataRetrievalService.cs:13]: Commented code generally means that it's not going to be used, and should thus not be present in the delivered code.

File: API/Services/DataRetrievalService.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+using System.Threading.Tasks;
+using Backend.API.Data;
+
+namespace Backend.API.Services
+{
+    public interface IDataRetrievalService
+    {
+        Task<Todo> GetTodo(int todoId);
+
+        Task<Todo[]> GetTodos();
```
Comment (FredrikAugust) [inline API/Services/DataRetrievalService.cs:13]:
```suggestion
        Task<Todo> GetTodo(int todoId);

        Task<Todo[]> GetTodos();
```

File: API/Services/DataRetrievalService.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+            HttpResponseMessage response =
+                await HttpClient.GetAsync($"https://jsonplaceholder.typicode.com/todos/{todoId}");
+            return await response.Content.ReadFromJsonAsync<Todo>();
+        }
+
+        public async Task<Todo[]> GetTodos()
+        {
+            HttpResponseMessage response = await HttpClient.GetAsync("https://jsonplaceholder.typicode.com/posts");
+            return await response.Content.ReadFromJsonAsync<Todo[]>();
+        }
```
Comment (FredrikAugust) [inline API/Services/DataRetrievalService.cs:39]:
```suggestion
```

Replace with your own data service. See comment further down.

File: API/Schemas/TodoSchema.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using Backend.API.Queries;
+using Backend.API.Services;
+
+namespace Backend.API.Schemas
+{
+    public class TodoSchema : GraphQL.Types.Schema
```
Comment (FredrikAugust) [inline API/Schemas/TodoSchema.cs:7]: Why do we have two schemas? Presuming we're only delivering one service to the frontend, I believe it's more reasonable to create a `RootSchema` and rather use queries for segregation.

File: API/Schemas/ForecastSchema.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using Backend.API.Queries;
+using Backend.API.Services;
+
+namespace Backend.API.Schemas
+{
+    public class ForecastSchema : GraphQL.Types.Schema
+    {
+        public ForecastSchema(IForecastDataRetrievalService forecastDataRetrievalService, IServiceProvider provider) : base(provider)
```
Comment (FredrikAugust) [inline API/Schemas/ForecastSchema.cs:9]: See comment under.

File: API/Queries/TodoQuery.cs [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+            Field<TodoType>("todo", arguments: new QueryArguments()
+            {
+                new QueryArgument<IntGraphType>
+                    {Name = "id", Description = "The ID of the todo to retrieve.", DefaultValue = 1}
+            }, resolve: context => dataRetrievalService.GetTodo(context.GetArgument<int>("id")));
+
+            Field<ListGraphType<TodoType>>("todos", resolve: _ => dataRetrievalService.GetTodos());
+        }
+    }
+}
```
Comment (FredrikAugust) [inline API/Queries/TodoQuery.cs:21]: Delete all todo code.

File: API/Queries/RootQuery.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    /*
+     * template for making a new Query
+     */
+    {
+        public RootQuery()
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
```
Comment (FredrikAugust) [inline API/Queries/RootQuery.cs:16]: This can be deduced from the other queries or Google'd.

```suggestion
```

Comment (FredrikAugust) [reply API/Queries/RootQuery.cs:16]: Suggestion: delete entire file.

File: API/Queries/ForecastQuery.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        public ForecastQuery(IForecastDataRetrievalService forecastDataRetrievalService)
+        {
+            Field<ForecastType>("forecast",description:"Latitude and Longitude defaults to the coordinates of Trondheim", arguments: new QueryArguments()
+                {
+                    new QueryArgument<FloatGraphType>(){Name = "lat", Description = "The lat (dec) of the forecast to retrieve.", DefaultValue = 63.446827},
+                    new QueryArgument<FloatGraphType>(){Name = "lon", Description = "The long (dec) of the forecast to retrieve.", DefaultValue = 10.421906}
+                },
+                // Two arguments that are used
+                // todo: Add support for decimals, only works with integers for now.
```
Comment (FredrikAugust) [inline API/Queries/ForecastQuery.cs:18]: Ref comment further down.

**Review by FredrikAugust — COMMENTED** [review]
Great work.

I can't request changes/approve the PR since I created it, but I included some comments.

I also agree that some of the files should be `SubClasses`, but at least personally I find it equally logical and legible to simply have no subfolders in the `MetAPI` directory, and have all files/classes thereunder.

File: API/Services/DataRetrievalService.cs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    public class DataRetrievalService : IDataRetrievalService
+    {
+        private HttpClient HttpClient { get; }
+
+        public DataRetrievalService()
+        {
+            HttpClient = new HttpClient();
+            HttpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            HttpClient.DefaultRequestHeaders.UserAgent.ParseAdd(
+                "Mozilla/5.0 (Macintosh; Intel Mac OS X x.y; rv:42.0) Gecko/20100101 Firefox/42.0");
```
Comment (FredrikAugust) [inline API/Services/DataRetrievalService.cs:25]: > All requests must (if possible) include an identifying User Agent-string (UA) in the request with the application/domain name, optionally version number. You should also include a company email address or a link to the company website where we can find contact information. If we cannot contact you in case of problems, you risk being blocked without warning.

The UserAgent should be updated to reflect this.

File: API/MetAPI/ForecastDataType.cs [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public ForecastDataType()
+        {
+            Field(data => data.Instant, false, typeof(InstantType)).Description("Details about current data"); 
+            Field(data => data.Next1Hours, false, typeof(Next1HoursType)).Description("Summary and details(inconsistent data available)");
+            Field(data => data.Next6Hours, false, typeof(Next6HoursType)).Description("Summary and details(inconsistent data available)");
+            Field(data => data.Next12Hours, false, typeof(Next12HoursType)).Description("Summary and details(inconsistent data available)");
+            /*
+             * Instant does only contain Details, while the others also have a summary
+             */
```
Comment (FredrikAugust) [inline API/MetAPI/ForecastDataType.cs:16]: In the future, this can be added as [class documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments).

Don't worry about this now!

File: API/Schemas/RootSchema.cs [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+using Backend.API.Queries;
+using Backend.API.Services;
+
+namespace Backend.API.Schemas
+{
+    public class RootSchema : GraphQL.Types.Schema
+    {
+        public RootSchema(IDataRetrievalService dataRetrievalService, IServiceProvider provider) : base(provider)
+        {
+            Query = new ForecastQuery(dataRetrievalService);
```
Comment (FredrikAugust) [inline API/Schemas/RootSchema.cs:11]: `ForecastQuery` should be `RootQuery` as it will contain all other sub-queries.

File: API/Services/DataRetrievalService.cs [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+using System.Threading.Tasks;
+using Backend.API.MetAPI;
+
+namespace Backend.API.Services
+{
+    public interface IDataRetrievalService
+    {
+        Task<Forecast>  GetForecast(float lat, float lon);
```
Comment (FredrikAugust) [inline API/Services/DataRetrievalService.cs:12]:
```suggestion
        Task<Forecast> GetForecast(float lat, float lon);
```

File: Backend.sln.DotSettings.user (hunk 10 above)
Comment (FredrikAugust) [reply Backend.sln.DotSettings.user:7]: This file should be deleted and added to `.gitignore`.

**Review by Kpaubert — COMMENTED** [review]
LGTM! 👍
